components/esp_https_server/src/https_server.c

```diff
@@ -213,51 +213,100 @@ static httpd_ssl_ctx_t *create_secure_context(const struct httpd_ssl_config *con
     }
 
 /* servercert = cert of server itself */
-    cfg->servercert_buf = (unsigned char *)malloc(config->servercert_len);
-    if (!cfg->servercert_buf) {
-        ESP_LOGE(TAG, "Could not allocate memory");
-        free((void *)cfg->cacert_buf);
+    if(config->servercert != NULL && config->servercert_len > 0) {
+        cfg->servercert_buf = (unsigned char *)malloc(config->servercert_len);
+
+        if(cfg->servercert_buf) {
+            memcpy((char *) cfg->servercert_buf, config->servercert, config->servercert_len);
+            cfg->servercert_bytes = config->servercert_len;
+        } else {
+            ESP_LOGE(TAG, "Could not allocate memory for server certificate");
+            free((void *) cfg->cacert_buf);
+            free(cfg);
+            free(ssl_ctx);
+            return NULL;
+        }
+    } else {
+#if defined(CONFIG_ESP_TLS_SERVER_SNI_HOOK)
+        if (config->sni_callback == NULL) {
+#endif
+        ESP_LOGE(TAG, "No Server certificate supplied");
+        free((void *) cfg->cacert_buf);
         free(cfg);
         free(ssl_ctx);
         return NULL;
+#if defined(CONFIG_ESP_TLS_SERVER_SNI_HOOK)
+        } else {
+            ESP_LOGW(TAG, "Server certificate not supplied, make sure to supply it in the SNI hook!");
+        }
+#endif
     }
-    memcpy((char *)cfg->servercert_buf, config->servercert, config->servercert_len);
-    cfg->servercert_bytes = config->servercert_len;
 
     /* Pass on secure element boolean */
     cfg->use_secure_element = config->use_secure_element;
     if (!cfg->use_secure_element) {
-        cfg->serverkey_buf = (unsigned char *)malloc(config->prvtkey_len);
-        if (!cfg->serverkey_buf) {
-            ESP_LOGE(TAG, "Could not allocate memory");
-            free((void *)cfg->servercert_buf);
-            free((void *)cfg->cacert_buf);
+        if (config->prvtkey_pem != NULL && config->prvtkey_len > 0) {
+            cfg->serverkey_buf = (unsigned char *) malloc(config->prvtkey_len);
+
+            if (cfg->serverkey_buf) {
+                memcpy((char *) cfg->serverkey_buf, config->prvtkey_pem, config->prvtkey_len);
+                cfg->serverkey_bytes = config->prvtkey_len;
+            } else {
+                ESP_LOGE(TAG, "Could not allocate memory for server key");
+                free((void *) cfg->servercert_buf);
+                free((void *) cfg->cacert_buf);
+                free(cfg);
+                free(ssl_ctx);
+                return NULL;
+            }
+        } else {
+#if defined(CONFIG_ESP_TLS_SERVER_SNI_HOOK)
+            if (config->sni_callback == NULL) {
+#endif
+            ESP_LOGE(TAG, "No Server key supplied");
+            free((void *) cfg->servercert_buf);
+            free((void *) cfg->cacert_buf);
             free(cfg);
             free(ssl_ctx);
             return NULL;
+#if defined(CONFIG_ESP_TLS_SERVER_SNI_HOOK)
+            } else {
+            ESP_LOGW(TAG, "Server key not supplied, make sure to supply it in the SNI hook!");
```

Comment:
I think instead of adding multiple `#if def` in the code.
We could merge the code when `SNI hook` is enabled into one and then keep the other code in another part.
I think the current format is affecting code readability.
The same is applicable to the above case.